File: src/ee/ads/MultiInterstitialAd.cpp
//
//  MultiInterstitialAd.cpp
//  ee_x
//
//  Created by Zinge on 10/11/17.
//
//

#include "ee/ads/MultiInterstitialAd.hpp"

#include <ee/core/ObserverHandle.hpp>
#include <ee/coroutine/Task.hpp>

namespace ee {
namespace ads {
using Self = MultiInterstitialAd;

Self::MultiInterstitialAd() {
    handle_ = std::make_unique<ObserverHandle>();
}

Self& Self::addItem(const std::shared_ptr<IInterstitialAd>& item) {
    items_.push_back(item);
    (*handle_).bind(*item).addObserver({
        .onLoaded =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });
    return *this;
}

void Self::destroy() {
    for (auto&& item : items_) {
        item->destroy();
    }
    items_.clear();
    handle_->clear();
}

bool Self::isLoaded() const {
    for (auto&& item : items_) {
        if (item->isLoaded()) {
            return true;
        }
    }
    return false;
}

Task<bool> Self::load() {
    bool result = false;
    for (auto&& item : items_) {
        if (item->isLoaded()) {
            continue;
        }
        if (co_await item->load()) {
            result = true;
        }
    }
    co_return result;
}

Task<bool> Self::show() {
    bool displayed = false;
    for (auto&& item : items_) {
        if (not displayed) {
            if (co_await item->show()) {
                displayed = true;
                continue;
            }
        }
        if (not item->isLoaded()) {
            item->load();
        }
    }
    co_return displayed;
}
} // namespace ads
} // namespace ee

File: src/unity/Runtime/Ads/MultiInterstitialAd.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EE {
    public class MultiInterstitialAd : ObserverManager<IInterstitialAdObserver>, IInterstitialAd {
        private readonly List<IInterstitialAd> _items;
        private readonly ObserverHandle _handle;

        public MultiInterstitialAd() {
            _items = new List<IInterstitialAd>();
            _handle = new ObserverHandle();
        }

        public MultiInterstitialAd AddItem(IInterstitialAd item) {
            _items.Add(item);
            _handle.Bind(item).AddObserver(new IInterstitialAdObserver {
                OnLoaded = () => DispatchEvent(observer => observer.OnLoaded?.Invoke()),
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
            return this;
        }

        public void Destroy() {
            foreach (var item in _items) {
                item.Destroy();
            }
            _items.Clear();
            _handle.Clear();
        }

        public bool IsLoaded => _items.Any(item => item.IsLoaded);

        public async Task<bool> Load() {
            var result = false;
            foreach (var item in _items) {
                if (item.IsLoaded) {
                    continue;
                }
                if (await item.Load()) {
                    result = true;
                }
            }
            return result;
        }

        public async Task<bool> Show() {
            var result = false;
            foreach (var item in _items) {
                if (!result) {
                    // Only process if there isn't any successfully displayed ad.
                    result = await item.Show();
                }
                if (!item.IsLoaded) {
                    // Load in background.
                    _ = item.Load();
                }
            }
            return result;
        }
    }
}